TA#72215 [16.0] [MIG] timesheet_holidays_time_control_compatible

rivo2302 opened the pull request.

rivo2302 committed:
[16.0] [MIG] timesheet_holidays_time_control_binding

lanto-razafindrabe commented on timesheet_holidays_time_control_binding/models/hr_leave.py:
```diff
@@ -0,0 +1,20 @@
+class HrLeave(models.Model):
+    _inherit = "hr.leave"
+
+    def action_validate(self):
```
![image](https://github.com/user-attachments/assets/803efc5b-2d92-4179-a7aa-533b7c7b123c)
Hello @rivo2302 
I have tested locally the module, I got this User Error of Odoo when trying to completely approving the time off. Odoo added a check in v16 that not allowed us to update created timesheet linked to a time off.

I have found unit test about it also failed.

And when testing locally, could you update screenshots please (not in debug mode) ? Thanks

lanto-razafindrabe committed:
Update README.rst

rivo2302 committed:
Merge branch '16.0' into TA#72215

rivo2302 committed:
[16.0][MIG]  timesheet_holidays_time_control_binding

lanto-razafindrabe commented on timesheet_holidays_time_control_binding/models/hr_leave.py:
```diff
@@ -0,0 +1,21 @@
+class HrLeave(models.Model):
+    _inherit = "hr.leave"
+
+    def action_validate(self):
```
hi @rivo2302 

I am curious, why did you not use the function in the module dependency `project_timesheet_holidays` : _timesheet_prepare_line_values() like in v14 ?

https://github.com/Numigi/odoo-timesheet-addons/blob/0f1495485d501f3a8cbcd7acfd030e9a17181cfc/timesheet_holidays_time_control_compatible/models/hr_leave.py#L12
It does not use context, but it passes.

rivo2302 replied:
Hello @lanto-razafindrabe ,

My bad, I didn't notice that the module name changed in V14. So, I followed the same logic from V12, which was using action_validate.

However, the difference in V16 is that they added a constraint preventing a timesheet line linked to a leave from being modified after its creation `def _check_can_update_timesheet(self)` . That's why the context is used.

I've sent back the module with its updated name.

lanto-razafindrabe replied:
Thank you @rivo2302 . So, next time, do not forget to do the checkout from 14 branch if the module exists, except for isidor module, to keep commit histories. In the other case, migrate from 12 😉

EDIT : I understand now, the module changed its name that's why you did not find it in 14.

rivo2302 committed:
[16.0][MIG]  timesheet_holidays_time_control_compatible

rivo2302 committed:
Merge branch '16.0' into TA#72215

rivo2302 committed:
[FIX] [precommit] main : manifest

lanto-razafindrabe committed:
Update README.rst

Restore it

Update test_validate_holiday_date.py

Update test_validate_holiday_date.py

Update test_validate_holiday_date.py

lanto-razafindrabe commented on timesheet_holidays_time_control_compatible/tests/test_validate_holiday_date.py:
```diff
@@ -0,0 +1,73 @@
+
+        leave2.sudo().action_validate()
+        for timesheet_line in leave2.sudo().timesheet_ids:
+            self.assertTrue(date_start <= timesheet_line.date_time <= date_end)
```
Hi @rivo2302  please keep the previous unit test in 14:
`            self.assertIn(
                timesheet_line.date_time, [date_start.date(), date_end.date()]
            )`
            
And improve this to test that the two days in start and end are dates of leave timesheets.

rivo2302 committed:
[16.0] [MIG] timesheet_holidays_time_control_compatible

lanto-razafindrabe committed:
TA#72215 [16.0] [CHG] timesheet_holidays_time_control_compatible : code quality
